src/pubsub/ua_pubsub_keystorage.c

```diff
@@ -430,4 +465,268 @@ UA_PubSubKeyStorage_detachKeyStorage(UA_Server *server, UA_PubSubKeyStorage *key
     }
 }
 
+/**
+ * @brief It holds the information required in the async callback to
+ * GetSecurityKeys method Call.
+ */
+typedef struct {
+    UA_Server *server;
+    UA_PubSubKeyStorage *ks;
+    UA_UInt32 startingTokenId;
+    UA_UInt32 requestedKeyCount;
+} sksClientContext;
+
+void
+sksClientCleanupCb(void *client, void *context);
+
+static void
+addDelayedSksClientCleanupCb(UA_Client *client){
+ UA_DelayedCallback *dc = (UA_DelayedCallback *)UA_calloc(1, sizeof(UA_DelayedCallback));
```

Comment:
Always check malloc results.
Often you can avoid malloc for delayed cleanups.
By adding the delayed callback structure to the object that shall be cleaned up.

we sometimes have embedded targets that run out of memory and still want to shut down gracefully.
So it’s good to require no memory on the cleanup path.